plugins/source/scheduler_round_robin.go

```diff
@@ -0,0 +1,197 @@
+	clientName := client.ID()
+	logger := p.logger.With().Str("table", table.Name).Str("client", clientName).Logger()
+
+	if parent == nil { // Log only for root tables, otherwise we spam too much.
+		logger.Info().Msg("top level table resolver started")
+	}
+	tableMetrics := p.metrics.TableClient[table.Name][clientName]
+
+	res := make(chan any)
+	go func() {
+		defer func() {
+			if err := recover(); err != nil {
+				stack := fmt.Sprintf("%s\n%s", err, string(debug.Stack()))
+				sentry.WithScope(func(scope *sentry.Scope) {
+					scope.SetTag("table", table.Name)
+					sentry.CurrentHub().CaptureMessage(stack)
+				})
+				logger.Error().Interface("error", err).Str("stack", stack).Msg("table resolver finished with panic")
+				atomic.AddUint64(&tableMetrics.Panics, 1)
+			}
+			close(res)
+		}()
+		if err := table.Resolver(ctx, client, parent, res); err != nil {
+			logger.Error().Err(err).Msg("table resolver finished with error")
+			atomic.AddUint64(&tableMetrics.Errors, 1)
+			return
+		}
+	}()
+
+	for r := range res {
+		p.resolveResourcesRoundRobin(ctx, table, client, parent, r, resolvedResources, depth)
+	}
+
+	// we don't need any waitgroups here because we are waiting for the channel to close
+	if parent == nil { // Log only for root tables and relations only after resolving is done, otherwise we spam per object instead of per table.
+		logger.Info().Uint64("resources", tableMetrics.Resources).Uint64("errors", tableMetrics.Errors).Msg("table sync finished")
+		p.logTablesMetrics(table.Relations, client)
+	}
+}
+
+func (p *Plugin) resolveResourcesRoundRobin(ctx context.Context, table *schema.Table, client schema.ClientMeta, parent *schema.Resource, resources any, resolvedResources chan<- *schema.Resource, depth int) {
+	resourcesSlice := helpers.InterfaceSlice(resources)
+	if len(resourcesSlice) == 0 {
+		return
+	}
+	resourcesChan := make(chan *schema.Resource, len(resourcesSlice))
+	go func() {
+		defer close(resourcesChan)
+		var wg sync.WaitGroup
+		for i := range resourcesSlice {
+			i := i
+			if err := p.resourceSem.Acquire(ctx, 1); err != nil {
+				p.logger.Warn().Err(err).Msg("failed to acquire semaphore. context cancelled")
+				wg.Wait()
+				// we have to continue emptying the channel to exit gracefully
+				return
+			}
+			wg.Add(1)
+			go func() {
+				defer p.resourceSem.Release(1)
+				defer wg.Done()
+				//nolint:all
+				resolvedResource := p.resolveResource(ctx, table, client, parent, resourcesSlice[i])
+				if resolvedResource == nil {
+					return
+				}
+				resourcesChan <- resolvedResource
+			}()
+		}
+		wg.Wait()
+	}()
+
+	var wg sync.WaitGroup
+	for resource := range resourcesChan {
+		resource := resource
+		resolvedResources <- resource
+		for _, relation := range resource.Table.Relations {
+			relation := relation
+			if err := p.tableSems[depth].Acquire(ctx, 1); err != nil {
+				// This means context was cancelled
```

Comment:
An opportunity to use `goto` (add a label in L196 where it `wg.Wait()`s and the fn ends) was missed here...